zkvm: design document (#273)
stellar/slingshot: zkvm/docs/zkvm-design.md

# ZkVM: fast, private, flexible blockchain contracts

[📄 ZkVM whitepaper](https://github.com/stellar/slingshot/files/3164078/zkvm.whitepaper.2019-05-09.pdf) (PDF)

## Table of contents

* [Introduction](#introduction)
* [Transactions](#transactions)
* [Program execution](#program-execution)
* [Contracts](#contracts)
* [Constraints](#constraints)
* [Scalability](#scalability)
* [Privacy](#privacy)
* [Performance](#performance)
* [Conclusion](#conclusion)
* [Future work](#future-work)
* [Authors](#authors-a-z)
* [References](#references)

## Introduction

In this paper, we present ZkVM, the _zero-knowledge virtual machine_: a transaction format for a shared, multi-asset, cryptographic ledger.

ZkVM aims at lowering the costs of participating in a global network by making the transactions _customizable_, _confidential_, highly _efficient_ and _simple_.

> **Customizable:** ZkVM is a safe and efficient environment for custom contracts. Contracts support realistic use cases via programmable constraints over encrypted data and assets. The VM supports a powerful language to express a wide range of higher-level protocols without the need to change the VM and upgrade the entire network.

Author: ```suggestion
**Customizable:** ZkVM permits custom contracts. Contracts support realistic use cases via programmable constraints over encrypted data and assets. The VM supports a powerful language to express a wide range of higher-level protocols without the need to change the VM and upgrade the entire network.
```

**Simple:** ZkVM requires no special compilers, debuggers, or optimizers. Its abstractions map well onto real-world concepts, and the technology stack consists of a small number of inventions reviewed by a global community of cryptographers and engineers.
ZkVM is an evolution of the authors’ prior work on [TxVM](https://chain.com/assets/txvm.pdf), which in turn was influenced by the design of [Bitcoin](https://bitcoin.org/bitcoin.pdf) and [Ethereum](https://ethereum.github.io/yellowpaper/paper.pdf). The motivation for such ledgers was originally expressed in the paper Secure Property Titles by [Nick Szabo](https://nakamotoinstitute.org/secure-property-titles/).

## Transactions

A ZkVM _transaction_ is a data structure around a _program_ that describes the issuance and transfer of _financial assets_.
A transaction also contains data such as protocol version, time bounds, and cryptographic proofs necessary to verify the correctness of the transaction.

![](zkvm-design-pics/zkvm-tx-chain.png)

> Assets are created via _issuance_: each asset is securely identified by its issuer’s predicate (public key), so only the issuer’s signature can authorize creation of additional units of the asset. Asset units cannot be destroyed by accident but must be formally _retired_ with a dedicated instruction.

Author: ```suggestion
Assets are created via _issuance_: each asset is securely identified by its issuer’s predicate (public key), so only the issuer’s signature can authorize creation of additional units of the asset (see [`issue`](zkvm-spec.md#issue)). Asset units are removed from circulation with a dedicated [`retire`](zkvm-spec.md#retire) instruction.
```

## Contracts

In ZkVM, _values_ are protected by _contracts_. Each contract contains a _payload_ and a _predicate_. The payload is an arbitrary number of items that could be _values_ or data parameters. The predicate is a condition that must be satisfied in order to unlock the items in the payload.

Transaction outputs are nothing more or less than contracts persisted to the blockchain state. During validation, contracts exist on the VM stack. ZkVM requires the transaction program to _clear_ contracts from the stack, which can only be done by satisfying their predicates. Once the items are unlocked, they can be used to form new items to be locked in either a transient contract or a persistent output.

![](zkvm-design-pics/zkvm-contract.png)

> The ZkVM instruction set is expressive enough to build a wide range of protocols using contracts: payment channels, order books, multi-factor/multi-party custody, collateralized loans, and even arbitrary state machines.

Author: yes, because the loops are unrolled by users manually into individual transactions (one tx per step).

> The ZkVM instruction set is expressive enough to build a wide range of protocols using contracts: payment channels, order books, multi-factor/multi-party custody, collateralized loans, and even arbitrary state machines.

Reviewer: Sorry to be unclear: it was a rhetorical question, meant to illustrate what readers might be wondering at this point.

> The ZkVM instruction set is expressive enough to build a wide range of protocols using contracts: payment channels, order books, multi-factor/multi-party custody, collateralized loans, and even arbitrary state machines.

Author: but it's still a good comment. Need to eventually clarify this.

## Performance

ZkVM does not permit loops or unbounded recursion. All immediately executed operations are very cheap, and the cost of execution is proportional to the length of the transaction program. A typical program consists of a handful of short instructions (`input`, `signtx`, `output` and one `cloak`), with 128 bytes of data for each input and 96 bytes for each output. If any contract reveals a sub-program, it must be encoded directly in the main program and cannot be constructed on the fly.

Verification of signatures, constraints, and Taproot commitments are relatively expensive because they involve the scalar multiplication of elliptic-curve points. For that reason, these operations are deferred and performed in a batch after the VM is finished executing the transaction’s program.

> Use of [Ristretto group](https://ristretto.group) and its state-of-the-art vectorized implementation in [Dalek Project](https://doc.dalek.rs/curve25519_dalek/) reduces the cost of one output verification to 1ms on a modern 3 GHz CPU, which implies a throughput of 1000 outputs per second per core. Verification is also highly parallelizable to optimize latency-critical deployments. Proving time is in the order of several milliseconds per output.

Reviewer: ```suggestion
Use of [the Ristretto group](https://ristretto.group) and its state-of-the-art vectorized implementation in [the Dalek project](https://doc.dalek.rs/curve25519_dalek/) reduces the cost of one output verification to 1ms on a modern 3 GHz CPU, which implies a throughput of 1000 outputs per second per core. Verification is also highly parallelizable to optimize latency-critical deployments. Proving time is in the order of several milliseconds per output.
```

Custom constraints created by contracts typically have significantly less overhead than the `cloak` instruction that merges and splits the input values into the outputs.

## Conclusion

ZkVM is a comprehensive solution that brings together the results of years of work by many researchers and developers in applied cryptography, programming languages, and cryptocurrencies.

It strikes the right balance between various partially conflicting requirements, producing a robust technology stack for the global financial system of the future.

> ZkVM is part of the [Slingshot open source project](../../), distributed under the Apache License, version 2.0.

Reviewer: ```suggestion
ZkVM is part of the [Slingshot open source project](https://github.com/stellar/slingshot), distributed under the Apache License, version 2.0.
```
